File: backend/embeddings.py
import os
import numpy as np
import faiss
from dotenv import load_dotenv
from langchain_text_splitters import RecursiveCharacterTextSplitter
from openai import OpenAI

load_dotenv()

_client = OpenAI(
    base_url=os.getenv("SUMOPOD_BASE_URL"),
    api_key=os.getenv("SUMOPOD_API_KEY"),
)

_CHUNK_SIZE = int(os.getenv("CHUNK_SIZE", "500"))
_CHUNK_OVERLAP = int(os.getenv("CHUNK_OVERLAP", "50"))
_EMBEDDING_MODEL = os.getenv("EMBEDDING_MODEL", "text-embedding-3-small")
_RETRIEVAL_TOP_K = int(os.getenv("RETRIEVAL_TOP_K", "4"))


def chunk_text(text: str) -> list[str]:
    splitter = RecursiveCharacterTextSplitter(
        chunk_size=_CHUNK_SIZE,
        chunk_overlap=_CHUNK_OVERLAP,
    )
    return splitter.split_text(text)


def _embed(texts: list[str]) -> np.ndarray:
    response = _client.embeddings.create(model=_EMBEDDING_MODEL, input=texts)
    vectors = [item.embedding for item in response.data]
    return np.array(vectors, dtype=np.float32)


def build_faiss_index(chunks: list[str]) -> tuple[faiss.IndexFlatL2, list[str]]:
    if not chunks:
        raise ValueError("Cannot build FAISS index from empty chunks list")

    embeddings = _embed(chunks)
    dimension = embeddings.shape[1]
    index = faiss.IndexFlatL2(dimension)
    index.add(embeddings)
    return index, chunks


def retrieve_chunks(
    query: str,
    index: faiss.IndexFlatL2,
    chunks: list[str],
    k: int = _RETRIEVAL_TOP_K,
) -> list[str]:
    query_vec = _embed([query])
    k = min(k, len(chunks))
    _, indices = index.search(query_vec, k)
    return [chunks[i] for i in indices[0] if i < len(chunks)]

File: api/lib/embeddings.ts
import { RecursiveCharacterTextSplitter } from '@langchain/textsplitters'
import { OpenAIEmbeddings } from '@langchain/openai'
import { MemoryVectorStore } from 'langchain/vectorstores/memory'

const CHUNK_SIZE    = parseInt(process.env.CHUNK_SIZE    ?? '500', 10)
const CHUNK_OVERLAP = parseInt(process.env.CHUNK_OVERLAP ?? '50',  10)
const TOP_K         = parseInt(process.env.RETRIEVAL_TOP_K ?? '4', 10)
const EMBED_MODEL   = process.env.EMBEDDING_MODEL ?? 'text-embedding-3-small'

function getEmbeddings(): OpenAIEmbeddings {
  return new OpenAIEmbeddings({
    model: EMBED_MODEL,
    apiKey: process.env.SUMOPOD_API_KEY ?? '',
    configuration: { baseURL: process.env.SUMOPOD_BASE_URL },
  })
}

export async function chunkText(text: string): Promise<string[]> {
  const splitter = new RecursiveCharacterTextSplitter({
    chunkSize: CHUNK_SIZE,
    chunkOverlap: CHUNK_OVERLAP,
  })
  return splitter.splitText(text)
}

export async function buildVectorStore(chunks: string[]): Promise<MemoryVectorStore> {
  if (!chunks.length) throw new Error('Cannot build vector store from empty chunks')
  return MemoryVectorStore.fromTexts(chunks, chunks.map(() => ({})), getEmbeddings())
}

export async function retrieveChunks(query: string, store: MemoryVectorStore): Promise<string[]> {
  const results = await store.similaritySearch(query, TOP_K)
  return results.map(doc => doc.pageContent)
}
